From 3db2a4ad1e561ef76280c2461d9cb6e9cdd4b787 Mon Sep 17 00:00:00 2001
From: tony lee <tony.lee@quantatw.com>
Date: Wed, 17 Apr 2019 18:43:40 +0800
Subject: [PATCH] Update period to d-bus which get from config

---
 manager.cpp | 7 ++++---
 manager.hpp | 2 +-
 2 files changed, 5 insertions(+), 4 deletions(-)

diff --git a/manager.cpp b/manager.cpp
index 0235b7b..b909d60 100644
--- a/manager.cpp
+++ b/manager.cpp
@@ -118,7 +118,7 @@ void Manager::driveLEDs(group& ledsAssert, group& ledsDeAssert)
             std::string objPath = std::string(PHY_LED_PATH) + it.name;
             log<level::DEBUG>("De-Asserting LED",
                               entry("NAME=%s", it.name.c_str()));
-            drivePhysicalLED(objPath, Layout::Action::Off, it.dutyOn);
+            drivePhysicalLED(objPath, Layout::Action::Off, it.dutyOn, it.period);
         }
     }
 
@@ -129,7 +129,7 @@ void Manager::driveLEDs(group& ledsAssert, group& ledsDeAssert)
             std::string objPath = std::string(PHY_LED_PATH) + it.name;
             log<level::DEBUG>("Asserting LED",
                               entry("NAME=%s", it.name.c_str()));
-            drivePhysicalLED(objPath, it.action, it.dutyOn);
+            drivePhysicalLED(objPath, it.action, it.dutyOn, it.period);
         }
     }
     return;
@@ -137,7 +137,7 @@ void Manager::driveLEDs(group& ledsAssert, group& ledsDeAssert)
 
 // Calls into driving physical LED post choosing the action
 void Manager::drivePhysicalLED(const std::string& objPath,
-                               Layout::Action action, uint8_t dutyOn)
+                               Layout::Action action, uint8_t dutyOn, uint16_t period)
 {
     using namespace phosphor::logging;
 
@@ -153,6 +153,7 @@ void Manager::drivePhysicalLED(const std::string& objPath,
     if (action == Layout::Action::Blink)
     {
         drivePhysicalLED(service->second, objPath, "DutyOn", dutyOn);
+        drivePhysicalLED(service->second, objPath, "Period", period);
     }
     drivePhysicalLED(service->second, objPath, "State",
                      getPhysicalAction(action));
diff --git a/manager.hpp b/manager.hpp
index 65f4c83..5f33c20 100644
--- a/manager.hpp
+++ b/manager.hpp
@@ -142,7 +142,7 @@ class Manager
      *  @param[in]  dutyOn    -  Duty Cycle ON percentage
      */
     void drivePhysicalLED(const std::string& objPath, Layout::Action action,
-                          uint8_t dutyOn);
+                          uint8_t dutyOn, uint16_t period);
 
     /** @brief Makes a dbus call to a passed in service name.
      *  This is now the physical LED controller
-- 
2.7.4

