From c1cc7629c98a4903c3a5d641191d4ac31332f9e5 Mon Sep 17 00:00:00 2001
From: tony lee <tony.lee@quantatw.com>
Date: Wed, 17 Apr 2019 18:36:58 +0800
Subject: [PATCH] Get parameter period in yaml config

---
 physical.cpp | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/physical.cpp b/physical.cpp
index e614b99..2401764 100644
--- a/physical.cpp
+++ b/physical.cpp
@@ -100,13 +100,14 @@ void Physical::stableStateOperation(Action action)
 void Physical::blinkOperation()
 {
     auto dutyOn = this->dutyOn();
+    auto periodMs = this->period();
+
+    led.setTrigger("timer");
 
     // Convert percent duty to milliseconds for sysfs interface
-    auto factor = periodMs / 100;
+    auto factor = periodMs / 100.0;
     led.setDelayOn(dutyOn * factor);
     led.setDelayOff((100 - dutyOn) * factor);
-
-    led.setTrigger("timer");
     return;
 }
 
-- 
2.7.4

