diff --git a/command/payload_cmds.cpp b/command/payload_cmds.cpp
index c32a510..d35778a 100644
--- a/command/payload_cmds.cpp
+++ b/command/payload_cmds.cpp
@@ -68,16 +68,22 @@ std::vector<uint8_t> activatePayload(const std::vector<uint8_t>& inPayload,
     handler.setChannelInSession();
 
     // Start the SOL payload
-    try
-    {
-        std::get<sol::Manager&>(singletonPool)
-            .startPayloadInstance(request->payloadInstance, handler.sessionID);
-    }
-    catch (std::exception& e)
-    {
-        log<level::ERR>(e.what());
-        response->completionCode = IPMI_CC_UNSPECIFIED_ERROR;
-        return outPayload;
+    int maxTries = 3;
+    for(int tries = 1;;tries++) {
+        try
+        {
+            std::get<sol::Manager&>(singletonPool)
+                .startPayloadInstance(request->payloadInstance, handler.sessionID);
+            break;
+        }
+        catch (std::exception& e)
+        {
+            log<level::ERR>(e.what());
+            if (tries == maxTries) {
+                response->completionCode = IPMI_CC_UNSPECIFIED_ERROR;
+                return outPayload;
+            }
+        }
     }
 
     response->inPayloadSize = endian::to_ipmi<uint16_t>(MAX_PAYLOAD_SIZE);
